external-authority-sources.md

```diff
@@ -0,0 +1,283 @@
+
+sample for an external source /server/api/integration/externalsources/orcid/entries?metadata=dc.contributor.author&query=Smith&size=2 
+```json
+{
+  "_embedded": {
+    "externalSourceEntries": [
+      {
+        "id": "Smith, Dean",
+        "display": "Smith, Dean",
+        "value": "Smith, Dean",
+        "metadata": {
+            "dc.identifier.orcid": "0000-0002-4271-0436",
+            "dc.identifier.uri": "https://orcid.org/0000-0002-4271-0436",
+            "dc.contributor.other": "University of Texas Southwestern Medical Center: TX, TX, US",
+            "person.familyName": "Smith",
+            "person.givenName": "Dean"
+        },
+        "type": "externalSource",
+        "_links": {
+          "authority": {
+            "href": "https://dspace7-internal.atmire.com/server/api/integration/authorities/authors/entryValues/d4b5ca88-9d6d-4a87-b905-fef0f8cae26c"
+          },
+          "self": {
+            "href": "https://dspace7-internal.atmire.com/server/api/integration/externalsources/orcid/entryValues/0000-0002-4271-0436"
+          }
+        }
+      },
+      {
+        "id": "Smith, Charles",
+        "display": "Smith, Charles",
+        "value": "Smith, Charles",
+        "metadata": {
+            "dc.identifier.orcid": "0000-0003-3681-2038",
+            "dc.identifier.uri": "https://orcid.org/0000-0003-3681-2038",
+            "dc.contributor.other": "University of Mississippi: University, MS, US",
+            "person.familyName": "Smith",
+            "person.givenName": "Charles"
+        },
+        "type": "externalSource",
+        "_links": {
+          "entity": {
+            "href": "https://dspace7-internal.atmire.com/server/api/core/item/6fd90bf5-b84f-47b3-aaec-a55bde3a2a5a"
+          },
+          "self": {
+            "href": "https://dspace7-internal.atmire.com/server/api/integration/externalsources/orcid/entryValues/0000-0003-3681-2038"
+          }
+        }
+      }
+    ]
+  }
+}
+```
+
+### single entry
+**GET /api/integration/externalsources/<:authority-name>/entryValues/<:entry-id>**
+
+It returns the data from one entry in an external source
+
+sample for an external source /api/integration/externalsources/orcid/entryValues/0000-0002-4271-0436 
+```json
+{
+  "id": "Smith, Dean",
+  "display": "Smith, Dean",
+  "value": "Smith, Dean",
+  "metadata": {
+      "dc.identifier.orcid": "0000-0002-4271-0436",
+      "dc.identifier.uri": "https://orcid.org/0000-0002-4271-0436",
+      "dc.contributor.other": "University of Texas Southwestern Medical Center: TX, TX, US",
+      "person.familyName": "Smith",
+      "person.givenName": "Dean"
+  },
+  "type": "externalSource",
+  "_links": {
+    "self": {
+      "href": "https://dspace7-internal.atmire.com/server/api/integration/externalsources/orcid/entryValues/0000-0002-4271-0436"
+    }
+  }
+}
+```
+
+**POST /api/integration/externalsources/<:authority-name>/entryValues/<:entry-id>/authority**
```

Comment:
We have discussed this integration internally and we arose to the point where we consider the "Import feature" should only be applicable to Entities. In the Authority context the authority value or key is stored alongside with it's related metadatavalue entry and, after the item is submitted/saved, on server side, an Event Consumer it's executed (https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/authority/indexer/AuthorityConsumer.java) to process it. You will only know what to import or to process at this point. So it doesn't make sense, to us to have here this POST method when we really need it's a PATCH to store the authority value and confidence for a certain metadatavalue https://github.com/DSpace/Rest7Contract/blob/master/metadata-patch.md).

Reply:
@paulo-graca This is actually part of the behavior which was discussed to be "problematic" or "confusing" in DSpace 6 and below.

Where in DSpace 6, you'd search on an ORCID author, and it somehow starts importing it into authority without the user understanding what happened, this is now a clear process.

The POST will create the authority value in solr (what used to be part of this consumer which worked with temp IDs). It will return the authority value.
The PATCH on the item metadata will add the returned authority value to the item.

For the user, this is what the import button will do: it will create the authority value and add it to the item. But these two parts got entangled deep in the code in previous versions of DSpace and are fixed here.

Reply:
The POST is not included here to replace the PATCH for metadata fields. The POST is included here to update solr. It should respond with the authority value, which is hereafter used in the item using the PATCH for metadata fields.

The reason for including this POST is to avoid using the `AuthorityConsumer`. The way the `AuthorityConsumer` works is creating a large amount of complex customizations to support ORCID directly from authority control. With this new external sources functionality, we don't need these customizations anymore and we can directly support REST access to convert an external source to the local authority.
I'd prefer not to extend the entire functionality underneath the `AuthorityConsumer` which was implemented for converting an ORCID ID to an authority record, and which would need to support any external source. It can be implemented much more cleanly with a separate POST. And this implementation will require a lot less work and is much less error prone

Reply:
@benbosman: While I understand that the `AuthorityConsumer` is not ideal in this scenario, I'm hesitant for the Entities Working Group to _change the behavior of_ the Authority backend, as that is outside the scope of our working group.  If we feel strongly that we *must* go this route, I feel we then *must* propose this to the broader DSpace 7 Working Group.

Overall, I still feel that having a `POST` generate an update to Solr is not an ideal approach in that it's not aligned with REST Best practices.  `POST` commands are supposed to create **resources** (usually objects). But, at this time, because of known flaws with how Authorities currently work, DSpace doesn't store an Authority "resource" in the database...instead DSpace stores the Authority in a metadata field, and links to an entry in a Solr index.

I agree that, longer term, we should rethink this approach to storing Authorities.  They should be managed in the database more formally. However, that is out-of-scope for DSpace 7...and I think this `POST` command only is "valid" once we fix the behavior in how Authorities are stored.

This is why I'm not sure this `POST` belongs in DSpace 7.  It could be (re)introduced in a future release (DSpace 8 perhaps) once we find time to redesign Authorities and/or merge Entities and Authorities (if plausible). But, until then, I think we should retain the current behavior of Authorities.   If this means that we cannot "import" into Authorities in the same way as we can with Entities, then we should re-scope our plans for DSpace 7 and perhaps (as Paulo suggested above) limit import to Entities only for now.